docs/syscall_descriptions_syntax.md

```diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
 define MY_PATH_MAX	PATH_MAX + 2
 ```
 
+## Conditional fields
+
+### In structures
+
+In syzlang, it's possible to specify a condition for every struct field that
+determines whether the field should be included or omitted:
+
+```
+header_fields {
+  magic       const[0xabcd, int16]
+  haveInteger int8
+} [packed]
+
+packet {
+  header  header_fields
+  integer int64  (if[value[header:haveField] == 0x1])
+  body    array[int8]
+} [packed]
+
+some_call(a ptr[in, packet])
+```
+
+In this example, the `packet` structure will include the field `integer` only
+if `header.haveInteger == 1`. That corresponds to e.g. the following program:
+```
+some_call(&AUTO={{AUTO, 0x1}, @value=0xabcd, []})
+```
+
+If `header.haveInteger` is not `1`, syzkaller will just pretend that the field
+`integer` does not exist -- much like it's done for `optional[T]`.
+```
+some_call(&AUTO={{AUTO, 0x0}, @void, []})
+```
+
+A structure that has conditional fields is assumed to be `[varlen]` and.  If
+a field of a `varlen` type or a conditional field appears in the middle of a
+structure, the structure must be `[packed].`
+
+Conditions on bitfields are prohibited:
+```
+struct {
+  f0 int
+  f1 int:3 (if[value[f0] == 0x1])  # It will not compile.
+}
+```
+
+But you may reference bitfields in your conditions:
+```
+struct {
+  f0 int:1
+  f1 int:7
+  f2 int   (if[value[f0] == value[f1]])
+} [packed]
+```
+
+### In unions
+
+Let's consider the following example.
+
+```
+struct {
+  type int
+  body alternatives
+}
+
+alternatives [
+  int     int64 (if[value[struct:type] == 0x1])
+  arr     array[int64, 5] (if[value[struct:type] == 0x1])
+  default int32
+] [varlen]
```

Comment:
You cannot mark structs `[varlen]` -- I've removed the notion from the document to make it less confusing. Structs must be marked with `[packed]` if they have a variable length field that is not the last field of the structure.

Variable length unions must be marked with `[varlen]`. I think it's done so that the fact is explicitly seen.